src/Microsoft.AspNet.TestHost/TestServer.cs

```diff
@@ -25,22 +25,32 @@ public class TestServer : IServerFactory, IDisposable
         private IDisposable _appInstance;
         private bool _disposed = false;
 
-        // REVIEW: we can configure services via AppStartup or via hostContext.Services
-        public TestServer(IConfiguration config, IServiceProvider serviceProvider, Action<IApplicationBuilder> configureApp, ConfigureServicesDelegate configureServices)
+        public TestServer(IServiceProvider serviceProvider, HostingContext context)
         {
-            var hostContext = new HostingContext()
+            if (context.ServerFactory == null)
             {
-                ApplicationName = "Test App",
-                Configuration = config,
-                ServerFactory = this,
-                StartupMethods = new StartupMethods(configureApp, configureServices)
-            };
-
-            _appInstance = new HostingEngine(serviceProvider).Start(hostContext);
+                context.ServerFactory = this;
+            }
+            _appInstance = new HostingEngine(serviceProvider, context.ApplicationName, context.ApplicationBasePath).Start(context);
         }
 
+        // REVIEW: we can configure services via AppStartup or via hostContext.Services
+        public TestServer(IConfiguration config, IServiceProvider serviceProvider, Action<IApplicationBuilder> configureApp, ConfigureServicesDelegate configureServices)
+            : this(serviceProvider, new HostingContext()
+                {
+                    ApplicationName = "Test App",
+                    Configuration = config,
+                    StartupMethods = new StartupMethods(configureApp, configureServices)
+                })
+        { }
+
         public Uri BaseAddress { get; set; } = new Uri("http://localhost/");
 
+        public static TestServer Create(HostingContext context)
```

Comment:
why does the test code now need to know `HostingContext`?  seems like moving the faked (but easily-created and understood) `IApplicationEnvironment` from test code into hosting has increased the concept count in code consuming these APIs.  would prefer to see just another `Create` method that takes the application name and path as parameters.